src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -256,22 +250,30 @@ private Task WriteFormattedInt32Async(int value, CancellationToken cancellationT
                 // CRLF for end of headers.
                 await WriteTwoBytesAsync((byte)'\r', (byte)'\n', cancellationToken).ConfigureAwait(false);
 
-                // Write body, if any
+                // Flush out anything buffered from the request headers to finish sending them.
+                await FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Because a server is unlikely to send a response until it's received the headers, so if the client is waiting for the server to send the first message, and all of our headers are buffered so the server doesn't know we've sent anything and thus won't send us anything, it'll deadlock unless we've flushed the buffer.

I understand the concern, and there are alternatives we could consider.  One is simply to say it's up to the request HttpContent to force the flush, e.g. in a stream like https://github.com/dotnet/corefx/pull/23180/files/f047e951e42c234a576d8bb9765bd0e6235084bf#diff-4d205133c92dc4f27822fa53f3c12653R238, overriding CopyToAsync and doing a FlushAsync on the destination stream as the very first thing.  Or we could do this flush, but not for certain cases we special-case, e.g. if the HttpContent is a ByteArrayContent (which would handle not only ByteArrayContent but also things derived from it, like StringContent and FormUrlEncodedContent), we could skip the flush and maybe even skip this mechanism altogether. (Note that the flush will also be necessary for Expect: 100-continue support, as if we don't flush out that header, the server won't know to send back status code 100, and we'll artifically delay the request by a second or whatever timeout we have for sending the request content... but we could also choose to only do the flush in that case.)